workflow/executor/executor.go

```diff
@@ -628,6 +641,153 @@ func (we *WorkflowExecutor) AddAnnotation(key, value string) error {
 	return common.AddPodAnnotation(we.ClientSet, we.PodName, we.Namespace, key, value)
 }
 
+type ConditionType string
+
+const (
+	ConditionTypeError   ConditionType = "error"
+	ConditionTypeWarning ConditionType = "warning"
+)
+
+func (we *WorkflowExecutor) EvaluateConditions(conditionMode ConditionType) error {
+
+	var resultsLocation *[]wfv1.ErrorCondition
+	var annotationKey string
+
+	if conditionMode == ConditionTypeError {
+		resultsLocation = &we.Template.Errors
+		annotationKey = common.AnnotationKeyErrors
+
+	} else if conditionMode == ConditionTypeWarning {
+		resultsLocation = &we.Template.Warnings
+		annotationKey = common.AnnotationKeyWarnings
+	} else {
+		return errors.InternalErrorf("The valid condition types are 'error' or 'warning', got %s instead", string(conditionMode))
+	}
+
+	results, err := we.evaluatePatternConditions(resultsLocation)
+	if err != nil {
+		return errors.InternalWrapError(err)
+	}
+
+	if results != nil {
+		errorResultBytes, err := json.Marshal(results)
+		if err != nil {
+			return errors.InternalWrapError(err)
+		}
+
+		return we.AddAnnotation(annotationKey, string(errorResultBytes))
+	}
+	return nil
+}
+
+func (we *WorkflowExecutor) fetchFileForErrorHandling(fileSource string) (logData []byte, err error) {
+
+	var logPath string
+
+	if fileSource[0] == '/' {
+		mainCtrID, err := we.GetMainContainerID()
+		if err != nil {
+			return nil, err
+		}
+		baseDir := "/argo/logs/"
+		uncompressedLogPath := baseDir + filepath.Base(fileSource)
+
+		// RuntimeExecutor.CopyFile gzips the file
+		logPath = uncompressedLogPath + ".gz"
+
+		if _, err := os.Stat(logPath); os.IsNotExist(err) {
+			err = os.MkdirAll(baseDir, os.ModePerm)
+			if err != nil {
+				err = errors.InternalWrapError(err)
+				return nil, err
+			}
+			err = we.RuntimeExecutor.CopyFile(mainCtrID, fileSource, logPath)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+	} else if fileSource == "stdout" {
+
+		logPath = "/argo/logs/main.log"
+		if _, err := os.Stat(logPath); os.IsNotExist(err) {
+			_, err = we.saveLogsToPath("/argo/logs", "main.log")
+			if err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		err = errors.InternalErrorf("fileSource must be an absolute path or 'stdout', got %s instead", fileSource)
+		return
+
+	}
+
+	logFile, err := os.Open(logPath)
+	if err != nil {
+		log.Errorf("Error attempting to open logfile %s", logPath)
+		return
+	}
+	defer logFile.Close()
+
+	if strings.HasSuffix(logPath, ".gz") {
+		decompressed, err := gzip.NewReader(logFile)
```

Comment:
I wonder if could generate performance issues on the controller. Are only file sources being decompressed or also stdout?